staging/src/k8s.io/legacy-cloud-providers/azure/azure_config.go

```diff
@@ -0,0 +1,123 @@
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	cloudConfigNamespace = "kube-system"
+	cloudConfigKey       = "cloud-config"
+)
+
+// The configure scope for Azure cloud provider secret. Supported values are:
+// * all            : configure applied for components (kubelet and controller-manager). This is the default value.
+// * node           : configure applied for nodes (kubelet).
+// * control-plane  : configure applied for control plane components (controller-manager).
+//
+// For different configure scope, the secret name would also be different:
+// * all            : secret name would be azure-cloud-provider.
+// * node           : secret name would azure-cloud-provider-node.
+// * control-plane  : secret name would be azure-cloud-provider-control-plane.
+type cloudConfigScope string
+
+const (
+	cloudConfigScopeAll          cloudConfigScope = "all"
+	cloudConfigScopeNode         cloudConfigScope = "node"
+	cloudConfigScopeControlPlane cloudConfigScope = "control-plane"
+)
+
+// The config type for Azure cloud provider secret. Supported values are:
+// * file   : The values are read from local cloud-config file.
+// * secret : The values from secret would override all configures from local cloud-config file.
+// * merge  : The values from secret would override only configurations that are explicitly set in the secret. This is the default value.
+type cloudConfigType string
+
+const (
+	cloudConfigTypeFile   cloudConfigType = "file"
+	cloudConfigTypeSecret cloudConfigType = "secret"
+	cloudConfigTypeMerge  cloudConfigType = "merge"
+)
+
+// InitializeCloudFromSecret initializes Azure cloud provider from Kubernetes secret.
+func (az *Cloud) InitializeCloudFromSecret() {
+	config, err := az.getConfigFromSecret()
+	if err != nil {
+		klog.Warningf("Failed to get cloud-config from secret: %v, skip initializing from secret", err)
+		return
+	}
+
+	if config == nil {
+		// Skip re-initialization if the config is not override.
+		return
+	}
+
+	if err := az.InitializeCloudFromConfig(config, true); err != nil {
+		klog.Errorf("Failed to initialize Azure cloud provider: %v", err)
+	}
+}
+
+func (az *Cloud) getConfigFromSecret() (*Config, error) {
+	// Read config from file and no override, return nil.
+	if az.Config.CloudConfigType == cloudConfigTypeFile {
+		return nil, nil
+	}
+
+	secretName := getConfigSecretName(az.Config.CloudConfigScope)
+	secret, err := az.kubeClient.CoreV1().Secrets(cloudConfigNamespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get secret %s: %v", secretName, err)
+	}
+
+	cloudConfigData, ok := secret.Data[cloudConfigKey]
+	if !ok {
+		return nil, fmt.Errorf("cloud-config is not set in the secret (%s)", secretName)
+	}
+
+	config := Config{}
+	if az.Config.CloudConfigType == "" || az.Config.CloudConfigType == cloudConfigTypeMerge {
+		// Merge cloud config, set default value to existing config.
```

Comment:
Might be missing something, how is this a merge?

Reply:
if some fields are already set in configure file, but not in secret, then it won't be override by another unmarshal. But if it's set in secret, then the value would be override.